lib/OpenQA/Client/Archive.pm

```diff
@@ -0,0 +1,245 @@
+        delete($job->{assets}->{repo});
+        $path->path('testresults/thumbnails')->make_path if $options->{'with-thumbnails'};
+        $self->download_assets;
+        $self->download_test_results;
+    }
+    else {
+        die "There's an error openQA client returned " . $res->code;
+    }
+}
+
+sub _download_test_result_details {
+    my ($self, $module) = @_;
+    #$self->client->max_response_size(10);
+    my $ua = $self->client;
+    if ($module->{screenshot}) {
+        #same structure is used for thumbnails and screenshots
+        my @dirnames = map { /md5_[^basename]/ ? $_ : () } keys %{$module};
+        # check if module has images
+        return 0 unless @dirnames;
+
+        my $dir = $module->{'md5_dirname'} || ($module->{'md5_1'} . '/' . $module->{'md5_2'});
+
+        $url->path("/image/$dir/" . $module->{md5_basename});
+
+        my $destination = $path->path('testresults/', $module->{screenshot});
+        my $tx = $ua->get($url)->res->content->asset->move_to($destination);
+
+        if ($options->{'with-thumbnails'}) {
+            $url->path('/image/', $dir, '/.thumbs/', $module->{md5_basename});
+            $ua->get($url)
+              ->res->content->asset->move_to($path->path('testresults/thumbnails/', $module->{md5_basename}));
+        }
+
+    }
+    elsif ($module->{text}) {
+        my $file = Mojo::File->new($path->path('testresults/' . $module->{text}));
+        $file->spurt($module->{text_data} // "No data\n");
+    }
+}
+
+sub _download_file_at {
+
+    my ($self, $file, $location) = @_;
+    my $ua = $self->client;
+
+    my $jobid = $job->{id};
+    $url->path("/tests/$jobid/file/$file");
+    $ua->on(start => \&_progress_monitior);
+    $ua->{filename} = $file;
+
+    my $tx = $ua->build_tx(GET => $url);
+    $ua->start($tx);
+
+    # fix the \r :)
+    print "\n";
+    my $destination = $location->path($file);
+    _download_handler($tx, $destination);
+}
+
+sub _download_handler {
+    my ($tx, $file) = @_;
+
+    my $filename = $file;
+    my $message;
+    $filename = $file->basename;
+
+    $message = {message => "Unexpected error while downloading $filename\n"} unless $tx;
+
+    if ($tx && $tx->error) {
+        my $err = $tx->error;
+        $message //= {message => "file not found.\n"} if $tx->error->{code} && $tx->error->{code} eq 404;
+        $message //= {message => "Unexpected error while downloading: @{ [$tx->error->{message}]}\n"}
+          if $tx->error->{message};
+    }
+    elsif ($tx->res->is_success && $tx->res->content->asset->move_to($file)) {
+        $message = {success => 1, message => "Asset $filename sucessfully downloaded and moved to $file\n"};
+    }
+    else {
+        warn $tx->req->url;
+        $message //= {message => "Unexpected error while moving $file.\n"};
+    }
+
+    # Everything is an error, except when there is sucess
+    print $message->{message} if $message->{success};
+    print $message->{message} unless $message->{success};
```

Comment:
That does not seem to make any sense. Both if and unless with the same?

Reply:
Code used to be print $message->{message} if $@ if i remember correctly :) Good catch!

Reply:
Fixed